jdaviz/configs/cubeviz/plugins/viewers.py

```diff
@@ -197,6 +198,58 @@ def __init__(self, *args, **kwargs):
         # default_tool_priority=['jdaviz:selectslice']
         super().__init__(*args, **kwargs)
 
+    def _get_spatial_subset_layers(self):
+        return [ls for ls in self.state.layers
+                if isinstance(getattr(ls.layer, 'subset_state', None), RoiSubsetState)]
+
+    def _get_spectral_subset_layers(self):
+        return [ls for ls in self.state.layers
+                if isinstance(getattr(ls.layer, 'subset_state', None), RangeSubsetState)]
+
+    def _get_marks_for_layers(self, layers):
+        layers_list = list(self.state.layers)
+        # here we'll assume that all custom marks are subclasses of Lines/GL but don't directly
+        # use Lines/LinesGL (so an isinstance check won't be sufficient here)
+        layer_marks = [m for m in self.figure.marks if m.__class__.__name__ in ['Lines', 'LinesGL']]
+        # and now we'll assume that the marks are in the same order as the layers, this should
+        # be the case as long as the order isn't manually resorted
+        return [layer_marks[layers_list.index(layer)] for layer in layers]
+
+    def _on_subset_delete(self, msg):
+        # delete any ShadowSpatialSpectral mark for which either of the spectral or spatial marks
+        # no longer exists
+        spectral_marks = self._get_marks_for_layers(self._get_spectral_subset_layers())
+        spatial_marks = self._get_marks_for_layers(self._get_spatial_subset_layers())
```

Comment:
Or instead of my comment above, maybe just have one method that return both at once. Then you only need to run the loop once instead of twice.

Reply:
in other places in the code, I only need one or the other.  So I could make a single method which does the loop and appends to create two lists and returns both of them, and then just ignore the one I don't need in the other cases.  Do you think that would be better?  Or did you have something else in mind?

Reply:
Ah, if you don't always need both, then you can ignore this. I don't see we ever have enough Subsets to make the loop laggy, so doesn't really matter.